File: AddOperator.java
// Time Complexity :  O(3^N)
// Space Complexity : O(N) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

import java.util.*;

class AddOperator {
    List<String> result;
    public List<String> addOperators(String num, int target) {
        result = new ArrayList<>();
        if (num == null || num.length() == 0) return result;
        helper(num, 0,0,0,"",target);
        return result;
    }

    private void helper(String num, int pivot, long calc, long tail, String path, int target){
        // base
        if (pivot == num.length()){
            if (calc == target){
                result.add(path);
            }
            return;
        }

        //logic
        for (int i=pivot;i<num.length();i++){
            if (num.charAt(pivot) == '0' && pivot != i) continue;
            long curr = Long.parseLong(num.substring(pivot,i+1));
            if (pivot == 0){
                helper(num, i+1, curr, curr, path+curr, target );
            } else {
                //+
                helper(num, i+1, calc+curr, +curr, path+"+"+curr, target );
                //-
                helper(num, i+1, calc-curr, -curr, path+"-"+curr, target );
                //*
                helper(num, i+1, calc-tail + (tail*curr), tail*curr, path+"*"+curr, target );
            }
        }
    }
}

File: add_operator.py
class Solution:
    result = ""
    def addOperators(self, num: str, target: int) -> List[str]:

        if len(num) == 0 or num == None:
            return result

        self.result = []
        self.helper(num , 0 , target , 0,0,"")

        return self.result
    
    def helper(self , num , index , target , calc , tail , path):

        #base
        if index == len(num):
            if calc == target:
                self.result.append(path)
            
            return

        #logic
        for i in range(index,len(num)):
            if num[index] == '0' and index != i:
                continue
            current = num[index:i+1]
            curr = int(current)
            if index == 0:
                self.helper(num , i + 1 , target , curr , curr ,path + str(curr) )
            
            else:
                # + case
                self.helper(num , i + 1 , target , calc + curr , +curr , path + "+" + str(curr))

                # - case
                self.helper(num , i + 1 , target , calc - curr , -curr , path + "-" + str(curr))

                # * case

                self.helper(num , i+1 , target , calc - tail + tail * curr , tail * curr , path +"*"+str(curr))
            
            #backtracking step

           
            

